### PR TITLE
Transforms

### PR DESCRIPTION
**Pull request by morganchen12**
Add support for defining transformations (i.e. sort) on the data from FirebaseDatabase before it's presented in a table/collection view. Still missing a few things. Addresses #183, depends on ~~#184~~. Also allows for consumers to create their own collections.

- [x] ~~Doc comments~~
- [x] ~~DRY code in collections~~
- [x] ~~Tests~~

@mcdonamp please review

### CODE REVIEW

**Comment by morganchen12** [comment]
cc @puf, who may have opinions on this

File: FirebaseDatabaseUI/FUICollection.h [hunk 1]
```diff
@@ -18,57 +18,109 @@
...
  */
-@protocol FUIArrayDelegate<NSObject>
+@protocol FUICollectionDelegate<NSObject>
 
 @optional
 
 /**
+ * Called before any other events are sent.
+ */
+- (void)arrayDidBeginUpdates:(id<FUICollection>)collection;
```
Comment (puf) [inline FirebaseDatabaseUI/FUICollection.h]: I intentionally did not add a "begin updates" event, but only a "we're done with all updates". What do you think you would catch in this method that you would catch with a simple "is this the first update?" check?

Comment (morganchen12) [inline FirebaseDatabaseUI/FUICollection.h]: This is just to better mirror the UIKit aggregation methods of [`beginUpdates`](https://developer.apple.com/reference/uikit/uitableview/1614908-beginupdates) and [`endUpdates`](https://developer.apple.com/reference/uikit/uitableview/1614890-endupdates).

Comment (morganchen12) [inline FirebaseDatabaseUI/FUICollection.h]: I suspect it's ok to drop the `beginUpdates` event if the platform doesn't call for it. :)

**Pull request merged**